File: go/113.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var ret [][]int
var cur []int

func helper(t *TreeNode, s, target int) {
	if t == nil {
		return
	}
	s += t.Val
	cur = append(cur, t.Val)
	if s == target && t.Left == t.Right {
		// I assume that,
		// there will not be a condition that left == right != nil, haha
		var tmp []int = make([]int, len(cur))
		copy(tmp, cur)
		ret = append(ret, tmp)
	}
	helper(t.Left, s, target)
	helper(t.Right, s, target)
	cur = cur[:len(cur)-1]
}

func pathSum(root *TreeNode, targetSum int) [][]int {
	ret = make([][]int, 0)
	cur = make([]int, 0)
	helper(root, 0, targetSum)
	return ret
}

File: cpp/113.cpp
#include <iostream>
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  void helper(TreeNode *t, int curSum, vector<int> &v) {
    if (!t) // if current node is null
      return;
    int sum = curSum + t->val;
    v.push_back(t->val);
    if (sum == target && !t->left && !t->right) { // sum matched, and t is leaf
      ret.push_back(vector<int>());
      ret.back().assign(v.begin(), v.end());
      v.pop_back();
      return;
    }
    // recursivly check
    helper(t->left, sum, v);
    helper(t->right, sum, v);

    // remove current value to back-tracking
    v.pop_back();
  }
  vector<vector<int>> ret;
  int target;
  vector<vector<int>> pathSum(TreeNode *root, int targetSum) {
    target = targetSum;
    vector<int> vec = vector<int>();
    helper(root, 0, vec);
    return ret;
  }
};
